File: observer-pattern/js/src/ConcreteObserver2.js
import { ObserverAPI } from './ObserverAPI.js'

// 具体的观察者实现类，也可以看成订阅者，关联对应的主题类。
// 不同的观察者可以对应不同的主题。
export class ConcreteObserver2 extends ObserverAPI {
  // 这里没有在构造器就绑定某个主题，而是从客户角度去注册观察者
  // 观察者发出更新通知，观察者自行监听
  update(content) {
    console.log(`${this.constructor.name}::update() [content = ${content}]`)
  }
}

File: observer-pattern/go/src/ConcreteObserver2.go
package src

import "fmt"

// 具体的观察者实现类2，也可以看成订阅者，关联对应的主题类。
type ConcreteObserver2 struct {
  name    string
  subject Subject
}

func (c *ConcreteObserver2) SetSubject(subject Subject) {
  c.subject = subject
}

func (c *ConcreteObserver2) SetName(name string) {
  c.name = name
}

func (c *ConcreteObserver2) GetName() string {
  return c.name
}

func (c *ConcreteObserver2) GetSubjectName() string {
  if c.subject != nil {
    return c.subject.GetName()
  }
  return "unknow"
}

// 发布者(主题)发出更新通知后，观察者(订阅者)自行监听和实现动作
func (c *ConcreteObserver2) Update(content string) {
  fmt.Println("ConcreteObserver2::Update() [subject.name = " + c.GetSubjectName() + " content = " + content + "]")
}
